In [54]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [55]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id",
    encoding="utf-8")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [56]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [57]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

In [58]:
df_market_data.columns

Index(['price_change_percentage_24h', 'price_change_percentage_7d',
       'price_change_percentage_14d', 'price_change_percentage_30d',
       'price_change_percentage_60d', 'price_change_percentage_200d',
       'price_change_percentage_1y'],
      dtype='object')

### Prepare the Data

In [59]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
df_market_data_scaled = StandardScaler().fit_transform(
    df_market_data[['price_change_percentage_24h', 
                    'price_change_percentage_7d',
                    'price_change_percentage_14d',
                    'price_change_percentage_30d',
                    'price_change_percentage_60d',
                    'price_change_percentage_200d',
                    'price_change_percentage_1y']]
)

df_market_data_scaled[:1]

array([[ 0.50852937,  0.49319307,  0.77220043,  0.23545963, -0.0674951 ,
        -0.35595348, -0.25163688]])

In [60]:
# Create a DataFrame with the scaled data
df_market_data_scaled_df = pd.DataFrame(df_market_data_scaled, columns=[
                    'price_change_percentage_24h', 
                    'price_change_percentage_7d',
                    'price_change_percentage_14d',
                    'price_change_percentage_30d',
                    'price_change_percentage_60d',
                    'price_change_percentage_200d',
                    'price_change_percentage_1y'
])

# Copy the crypto names from the original data
# Set the coinid column as index
df_market_data_scaled_df.index = df_market_data.index

# Display sample data
df_market_data_scaled_df.sample(5)


price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
chainlink                        0.011397                    2.572251   
leo-token                        0.051758                   -0.928381   
bitcoin-cash                     1.193036                    2.000959   
omisego                          1.919812                    0.370447   
ontology                        -0.409044                   -0.906963   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
chainlink                        1.101647                    -0.490495   
leo-token                       -0.871918                     0.058782   
bitcoin-cash                     1.760610                     0.545842   
omisego                         -1.619761                    -0.409716   
ontology                        -1.298986                    -1.393153   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
chainlink                       -0.931954                      0.387759   
leo-token                       -0.159250                     -0.512071   
bitcoin-cash                    -0.291203                     -0.499848   
omisego                          1.696480                      0.811207   
ontology                        -0.696937                     -0.422835   

              price_change_percentage_1y  
coin_id                                   
chainlink                      -0.018284  
leo-token                      -0.265036  
bitcoin-cash                   -0.270317  
omisego                        -0.021888  
ontology                       -0.292344

---

### Find the Best Value for k Using the Original Data.

In [61]:
# Create a list with the number of k-values from 1 to 11 
# was a little unclear if this wanted 10 or 11 total values. Went with the code below for a total of 10 k-values. 
k = list(range(1, 11))

In [62]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list

for i in k:
    k_model = KMeans(n_clusters=i, random_state=0)
    k_model.fit(df_market_data_scaled)
    inertia.append(k_model.inertia_)


c:\Users\krist\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
c:\Users\krist\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\krist\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
c:\Users\krist\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environme

In [63]:
# Create a dictionary with the data to plot the Elbow curve
# Create a DataFrame with the data to plot the Elbow curve

elbow_data = {"k": k, "inertia": inertia}
df_elbow = pd.DataFrame(elbow_data)
df_elbow.head()

k     inertia
0  1  287.000000
1  2  195.820218
2  3  123.190482
3  4   79.022435
4  5   65.302379

In [64]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.

elbow_chart_1 = df_elbow.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve", 
    xticks=k,
    label='Elbow Curve'
)

display(elbow_chart_1)

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** The best value for 'k' appears to lie somewhere between 3 and 4. The first apparent elbow is present at 3, though another is visable at 4. I would begin my analysis with a clustering of 4.

When completing the graph below with a cluster of 4, you see a couple of potential outliers, and two distinct clusters. When using 3, you have two outliers and one main cluster. 

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [65]:
# Initialize the K-Means model using the best value for k# Define the model with optimal number of clusters
model = KMeans(n_clusters=4, random_state=0)

In [66]:
# Fit the K-Means model using the scaled data
model.fit(df_market_data_scaled)


c:\Users\krist\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
c:\Users\krist\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4, random_state=0)

In [67]:
# Predict the clusters to group the cryptocurrencies using the scaled data
kmeans_predictions = model.predict(df_market_data_scaled)

# Print the resulting array of cluster values.
kmeans_predictions



array([0, 0, 3, 3, 0, 0, 0, 0, 0, 3, 3, 3, 3, 0, 3, 0, 3, 3, 0, 3, 3, 0,
       3, 3, 3, 3, 3, 3, 0, 3, 3, 3, 1, 0, 3, 3, 2, 3, 3, 3, 3])

In [68]:
# Create a copy of the DataFrame
df_market_data_predictions = df_market_data_scaled_df.copy()

df_market_data_predictions.sample(5)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                     
crypto-com-chain                     0.331280                   -1.614844   
chainlink                            0.011397                    2.572251   
iota                                 0.259097                    0.249508   
ripple                              -0.040764                   -0.810928   
dash                                -0.296940                    0.094763   

                  price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                      
crypto-com-chain                    -1.054521                    -0.729931   
chainlink                            1.101647                    -0.490495   
iota                                -0.478953                    -0.218997   
ripple                               0.249458                    -0.050388   
dash                                 0.040040                    -0.358830   

                  price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                       
crypto-com-chain                    -0.350155                     -0.022866   
chainlink                           -0.931954                      0.387759   
iota                                -0.735815                     -0.329179   
ripple                              -0.373164                     -0.458259   
dash                                -0.558527                     -0.535732   

                  price_change_percentage_1y  
coin_id                                       
crypto-com-chain                   -0.034570  
chainlink                          -0.018284  
iota                               -0.285310  
ripple                             -0.295546  
dash                               -0.284071

In [69]:
# Add a new column to the DataFrame with the predicted clusters
df_market_data_predictions["clusters"] = kmeans_predictions

# Display sample data
df_market_data_predictions.sample(5)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                     
theta-token                         -1.612188                   -1.682027   
binancecoin                          0.891871                    1.327295   
dash                                -0.296940                    0.094763   
stellar                             -0.277543                   -0.385209   
crypto-com-chain                     0.331280                   -1.614844   

                  price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                      
theta-token                         -0.816921                     1.148607   
binancecoin                          0.800214                    -0.057148   
dash                                 0.040040                    -0.358830   
stellar                             -0.153243                    -0.371816   
crypto-com-chain                    -1.054521                    -0.729931   

                  price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                       
theta-token                          1.712641                      1.502992   
binancecoin                          0.778653                     -0.188232   
dash                                -0.558527                     -0.535732   
stellar                             -0.656403                     -0.353387   
crypto-com-chain                    -0.350155                     -0.022866   

                  price_change_percentage_1y  clusters  
coin_id                                                 
theta-token                         0.286977         3  
binancecoin                        -0.225533         0  
dash                               -0.284071         3  
stellar                            -0.270874         3  
crypto-com-chain                   -0.034570         3

In [88]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.

scatter_1 = df_market_data_predictions.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    hover_cols='coin_id',
    by="clusters", 
    label="Clusters"
)

display(scatter_1)

:NdOverlay   [clusters]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [89]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)


In [90]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
df_market_data_scaled_pca = pca.fit_transform(df_market_data_scaled_df)

# View the first five rows of the DataFrame. 
df_market_data_scaled_pca[:5]


array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [91]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
exp_variance = pca.explained_variance_ratio_

sum(exp_variance)


0.895031657030984

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** The total explained variance of the three principal components is 0.895031657030984

In [92]:
# Create a new DataFrame with the PCA data.
market_data_pca_df = pd.DataFrame(df_market_data_scaled_pca, columns=["PCA1", "PCA2", "PCA3"])

# Copy the crypto names from the original data
# Set the coinid column as index
market_data_pca_df.index = df_market_data.index

# Display sample data
market_data_pca_df.sample(10)


PCA1      PCA2      PCA3
coin_id                                       
cardano          -0.345600  0.729439  1.478013
litecoin         -0.649468  0.432165  0.600303
ethlend           8.089018 -3.896891  2.301382
tron             -0.393352 -0.108192 -0.012756
bitcoin-cash     -1.157800  2.041209  1.859715
tether           -0.433070 -0.168126 -0.641752
cosmos           -0.306272  0.703415  1.714224
neo              -0.362120 -0.986914 -0.728752
crypto-com-chain -0.248198 -1.376252 -1.462026
okb               0.064075 -1.269825 -1.098829

---

### Find the Best Value for k Using the PCA Data

In [93]:
# Create a list with the number of k-values from 1 to 11
pca_k = list(range(1, 11))

In [94]:
# Create an empty list to store the inertia values
pca_inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list

for i in k:
    pca_k_model = KMeans(n_clusters=i, random_state=0)
    pca_k_model.fit(market_data_pca_df)
    pca_inertia.append(pca_k_model.inertia_)

c:\Users\krist\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
c:\Users\krist\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\krist\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
c:\Users\krist\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environme

In [95]:
# Create a dictionary with the data to plot the Elbow curve
# Create a DataFrame with the data to plot the Elbow curve
pca_elbow_data = {"k": pca_k, "inertia": pca_inertia}
pca_df_elbow = pd.DataFrame(pca_elbow_data)

pca_df_elbow

k     inertia
0   1  256.874086
1   2  165.901994
2   3   93.774626
3   4   49.665497
4   5   37.878747
5   6   27.618972
6   7   21.182776
7   8   17.389823
8   9   13.593650
9  10   10.559358

In [96]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow_chart_2 = pca_df_elbow.hvplot.line(
    x="k", 
    y="inertia", 
    title="PCA Elbow Curve", 
    xticks=k, 
    label="PCA Elbow Curve"
)

display(elbow_chart_2)

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** It appears as though the best value for K when using the PCA data is 4. The 


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** 

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [97]:
# Initialize the K-Means model using the best value for k
pca_model = KMeans(n_clusters=4, random_state=0)

In [98]:
# Fit the K-Means model using the PCA data
pca_model.fit(market_data_pca_df)

c:\Users\krist\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
c:\Users\krist\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4, random_state=0)

In [99]:
# Predict the clusters to group the cryptocurrencies using the PCA data

pca_predictions = pca_model.predict(market_data_pca_df)
# Print the resulting array of cluster values.
pca_predictions

array([0, 0, 3, 3, 0, 0, 0, 0, 0, 3, 3, 3, 3, 0, 3, 0, 3, 3, 0, 3, 3, 0,
       3, 3, 3, 3, 3, 3, 0, 3, 3, 3, 1, 0, 3, 3, 2, 3, 3, 3, 3])

In [100]:
# Create a copy of the DataFrame with the PCA data
market_data_pca_pred = market_data_pca_df.copy()

# Add a new column to the DataFrame with the predicted clusters
market_data_pca_pred['clusters'] = pca_predictions

# Display sample data
market_data_pca_pred.head(5)


PCA1      PCA2      PCA3  clusters
coin_id                                             
bitcoin      -0.600667  0.842760  0.461595         0
ethereum     -0.458261  0.458466  0.952877         0
tether       -0.433070 -0.168126 -0.641752         3
ripple       -0.471835 -0.222660 -0.479053         3
bitcoin-cash -1.157800  2.041209  1.859715         0

In [101]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.

scatter_2 = market_data_pca_pred.hvplot.scatter(
    x="PCA1",
    y="PCA2",
    hover_cols='coin_id',
    by="clusters",
    label='PCA Clusters'
)

display(scatter_2)

:NdOverlay   [clusters]
   :Scatter   [PCA1]   (PCA2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [102]:
# Composite plot to contrast the Elbow curves
composite_elbow = elbow_chart_1 * elbow_chart_2

display(composite_elbow)

:Overlay
   .Curve.Elbow_Curve     :Curve   [k]   (inertia)
   .Curve.PCA_Elbow_Curve :Curve   [k]   (inertia)

In [106]:
# Composite plot to contrast the clusters
composite_scatter = scatter_1 + scatter_2
composite_scatter_nested = scatter_1 * scatter_2

display(composite_scatter)

display(composite_scatter_nested)

:Layout
   .NdOverlay.Clusters     :NdOverlay   [clusters]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.PCA_Clusters :NdOverlay   [clusters]
      :Scatter   [PCA1]   (PCA2,coin_id)

:Overlay
   .NdOverlay.Clusters     :NdOverlay   [clusters]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.PCA_Clusters :NdOverlay   [clusters]
      :Scatter   [PCA1]   (PCA2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** In this case, the impact seems to be loss of information and an impact on cluster quality. When using few clusters, our graphs are simplier, but the outliers present seem to distort the data between outliers and non-outliers. In contrast, when moving from 3 to 4 clusters, the outliers are captured by their own cluster, allowing a more nuanced picture of the majority of the data points. It seems that in general, it may be easier to understand and explain the characteristics of the resulting clusters with smaller cluster amounts, as with fewer features, the complexity of the clustering structure may be reduced. Additionally, it likely lends itself to improved computational efficiency with massive datasets. I would argue that in the specific case at hand, increasing the clustering from 3 to 4 provides a much more meaningful insight allowing us to see cluster and segement the majority of the datapoints in the dataset.